### ISSUE TITLE
Intermittent compile failures using field interpolation

### ISSUE THREAD

**Issue by U007D** [dtolnay/thiserror#46]
I'm getting puzzling behavior when using field interpolation:
```rust
mod msg {
    pub const ERR_ARG_NOT_CONVERTIBLE_TO_UTF_8: &str =
        "Error: supplied command-line argument not convertible to UTF-8";
}

use thiserror::Error;

#[derive(Debug, Error, PartialEq)]
pub enum Error {
    #[error("{}: {value:?}", msg::ERR_ARG_NOT_CONVERTIBLE_TO_UTF_8)]
    ArgNotConvertibleToUtf8 { value: std::ffi::OsString },
}
```
yields:
```
error: there is no argument named `value`
  --> src/error.rs:15:18
   |
15 |     #[error("{}: {value:?}", msg::ERR_ARG_NOT_CONVERTIBLE_TO_UTF_8)]
   |                  ^^^^^^^^^
```

But simply by changing the `#[error()]` attribute to:
```rust
    #[error("{}: {:?}", msg::ERR_ARG_NOT_CONVERTIBLE_TO_UTF_8, value)]
```
everything works as expected.

Then I discovered something really puzzling: what's "intermittent" about this issue is that when trying to isolate it, I discovered that:
```rust
pub enum MyError {
    #[error("uh oh: {value:?}")]
    Whoops { value: std::ffi::OsString },
}
```
works perfectly fine.
  
To my eye, these are the same use case!  Any ideas why the `ArgNotConvertibleToUtf8` variant's interpolated `#[error()]` attribute is problematic?

Great crate, by the way.  I *love* the design principle that `thiserror` does not appear in my library's API.

**Comment by dtolnay** [comment]
It has to do with this item in the readme:

> When providing your own format args, the shorthand does not kick in

so if you provide one format arg, you need to provide them all (as you found).

I've been meaning to fix it so that code you tried (`"{}: {value:?}"`) would work, but it requires more sophisticated parsing of the format string and format args than the current approach and I haven't gotten to it yet. I would accept a fix though!

**Comment by U007D** [comment]
I had read that, but I guess I don't understand quite what is meant by "providing [my] own format args" --I assumed the only time I wouldn't be providing my own format args would be if the corresponding enum variant did not have a payload.

If it means, "if one uses interpolation with one brace, one must use interpolation for all of them", it turns out that I actually started with
```rust
#[error("{msg::ERR_ARG_NOT_CONVERTIBLE_TO_UTF_8}: {value:?})]
```
which gave
```
error: invalid format string: expected `'}'`, found `':'`
  --> src/error.rs:10:16
   |
10 |     #[error("{msg::ERR_ARG_NOT_CONVERTIBLE_TO_UTF_8}: {value:?}")]
   |              - ^ expected `}` in format string
   |              |
   |              because of this opening brace
   |
   = note: if you intended to print `{`, you can escape it using `{{`
```
Using the more conventional format is a workaround--it's just that I have been waiting for interpolated strings for *so long* in Rust (would ultimately like to see them accepted by `println!()` and friends)--I was excited to use them. :)

**Comment by dtolnay** [comment]
I don't intend to support that style. Arbitrary expressions shouldn't go in the format string. It is specifically for the 4 shorthands listed in the readme.

**Comment by U007D** [comment]
Which is fine.  So if
> When providing your own format args
does mean "if you provide one interpolated binding, you must provide them all", then the behavior I saw makes sense.

I will take a look at the parsing code over this holiday season, and if I can make a PR happen, I'd be happy to.

Thanks, David.

**Comment by dtolnay** [comment]
This was fixed in 1.0.7 so your original code should work now.
